pkg/promtail/targets/awss3target.go

```diff
@@ -0,0 +1,228 @@
+	}, []string{"path"})
+	s3ReadLines = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "s3_read_lines_total",
+		Help:      "Number of lines read.",
+	}, []string{"path"})
+	s3ObjectsActive = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "promtail",
+		Name:      "s3_files_active_total",
+		Help:      "Number of active files.",
+	})
+	s3LogLengthHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "promtail",
+		Name:      "s3_log_entries_bytes",
+		Help:      "the total count of bytes",
+		Buckets:   prometheus.ExponentialBuckets(16, 2, 8),
+	}, []string{"path"})
+)
+
+// Config describes behavior for Target
+type s3Config struct {
+	SyncPeriod time.Duration `yaml:"sync_period"`
+	Stdin      bool          `yaml:"stdin"`
+}
+
+// RegisterFlags register flags.
+func (cfg *s3Config) RegisterFlags(flags *flag.FlagSet) {
+	flags.DurationVar(&cfg.SyncPeriod, "target.sync-period", 10*time.Second, "Period to resync directories being watched and files being tailed.")
+	flags.BoolVar(&cfg.Stdin, "stdin", false, "Set to true to pipe logs to promtail.")
+}
+
+// S3Target describes S3 target
+type S3Target struct {
+	logger log.Logger
+
+	handler   api.EntryHandler
+	positions positions.Positions
+	labels    model.LabelSet
+
+	watch map[string]*objectReader
+	path  string
+	quit  chan struct{}
+	done  chan struct{}
+
+	config       aws.S3Config
+	objectClient chunk.ObjectClient
+	prefix       string
+	syncPeriod   time.Duration
+}
+
+// NewFileTarget create a new FileTarget.
+func NewS3Target(logger log.Logger, handler api.EntryHandler, positions positions.Positions, jobName string, objectClient chunk.ObjectClient, s3Config *scrape.S3Targetconfig) (*S3Target, error) {
+	// // object store sync period should not be less than 1 minute
+	// if s3Config.SyncPeriod < 1*time.Minute {
+	// 	return nil, errors.New("SyncPeriod should not be less than 1 minute")
+	// }
+
+	// add prefix label
+	labels := s3Config.Labels.Merge(model.LabelSet{
+		model.LabelName("s3_prefix"): model.LabelValue(s3Config.Prefix),
+	})
+
+	t := &S3Target{
+		logger:       logger,
+		handler:      handler,
+		positions:    positions,
+		quit:         make(chan struct{}),
+		done:         make(chan struct{}),
+		labels:       labels,
+		objectClient: objectClient,
+		watch:        map[string]*objectReader{},
+		prefix:       s3Config.Prefix,
+		syncPeriod:   s3Config.SyncPeriod,
+	}
+
+	go t.run()
+	return t, nil
+}
+
+func (t *S3Target) run() {
+	defer func() {
+		for _, v := range t.watch {
+			level.Info(t.logger).Log("msg", "updating object last position", "object", v.object.Key)
+			v.markPositionAndSize()
+			level.Info(t.logger).Log("msg", "stopping object reader", "object", v.object.Key)
+			v.stop()
+		}
+		close(t.done)
+	}()
+
+	ticker := time.NewTicker(t.syncPeriod)
+	for {
+		select {
+		case <-ticker.C:
+			objects, _, err := t.objectClient.List(context.Background(), t.prefix)
```

Comment:
I'm afraid this will hang on large buckets, but I think that's fine for an initial feature.

Reply:
Yep, you are right. We have to add multi part download for larger size files. Added a note here https://github.com/grafana/loki/blob/ab130b09eb965c3baa8f94d02c63b15704b1c300/docs/clients/promtail/configuration.md#s3_config